### PR TITLE
Cleanup+Speed-up DataStream Metadata handling

### PR DESCRIPTION
**Pull request by original-brownbear**
No need to copy maps so often, this was quite slow in many-shards
benchmarking which uses lots of datastreams.

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -1515,14 +1510,7 @@ public Builder removeIndexTemplate(String name) {
...
-            DataStreamMetadata dataStreamMetadata = (DataStreamMetadata) customs.get(DataStreamMetadata.TYPE);
-            if (dataStreamMetadata != null) {
-                return dataStreamMetadata.dataStreams().get(dataStreamName);
-            } else {
-                return null;
-            }
+            return dataStreams().dataStreams().get(dataStreamName);
         }
 
         public Builder dataStreams(Map<String, DataStream> dataStreams, Map<String, DataStreamAlias> dataStreamAliases) {
```
Comment (joegallo) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1516]: In light of my previous comment, I'm curious whether this is better with the current signature versus perhaps being named `dataStreamMetadata` and receiving a `DataStreamMetadata`.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 2]
```diff
@@ -1547,29 +1535,21 @@ public Builder put(DataStream dataStream) {
...
-            Map<String, DataStreamAlias> existingDataStreamAliases = Optional.ofNullable(
-                (DataStreamMetadata) this.customs.get(DataStreamMetadata.TYPE)
-            ).map(dsmd -> new HashMap<>(dsmd.getDataStreamAliases())).orElse(new HashMap<>());
-
-            this.customs.put(DataStreamMetadata.TYPE, new DataStreamMetadata(existingDataStreams, existingDataStreamAliases));
+            this.customs.put(DataStreamMetadata.TYPE, dataStreams().withAddedDatastream(dataStream));
             return this;
         }
 
+        private DataStreamMetadata dataStreams() {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Sure makes sense :)

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1516]: Yea lets do a follow-up for this IMO. Looking over this all these methods are needlessly complex and we can save tons of code here doing similar things like I did here already (as well as make the code a little more obviously correct in the process :)).

**Review by martijnvg — APPROVED** [review]
Thanks for making this better @original-brownbear @joegallo! LGTM 👍

**Pull request merged**